### PR TITLE
Initialization of bi-direction LSTM python code 

### PR DESCRIPTION
**Pull request by SahandFarhoodi**
The branch `master` should only contain codes that Shane and/or Frank have reviewed.
All developments must be done on branch `develop`, and then it can be merged to master via a PR.
There is a third branch called `work` which is used to develop code when a PR is still going on that tries to merge `develop` into `master`.

### CODE REVIEW

File: word_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+from keras.models import Sequential
+from keras.layers import LSTM
+from keras.layers import Dense
+from keras.layers import TimeDistributed
+from keras.layers import Bidirectional
+from keras.layers import Embedding
+from keras.layers import Dropout
+
+
+def is_all_english(str):
```
Comment (sffc) [inline word_segmentation.py:18]: More formally, what you want to check is if the string is ASCII.  See some solutions here:

https://stackoverflow.com/q/196345/1407170

File: word_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+        if ord(ch) > 130:
+            # print("character = {}, ord = {}".format(ch, ord(ch)))
+            return False
+    return True
+
+
+def sigmoid(x):
+    """
+    Computes the sigmoid function for a scalar
+    x: the scalar
```
Comment (sffc) [inline word_segmentation.py:40]: Suggestion: Here and elsewhere: Google style is to put a header before the list of arguments, like this:

```python
def sigmoid(x):
  """
  Computes the sigmoid function for a scalar.

  Args:
    x: the scalar
  """
```

File: word_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+from keras.layers import Dense
+from keras.layers import TimeDistributed
+from keras.layers import Bidirectional
+from keras.layers import Embedding
+from keras.layers import Dropout
+
+
+def is_all_english(str):
+    """
+    A very basic function that checks if all elements of str are english characters or common expressions
```
Comment (sffc) [inline word_segmentation.py:20]: Optional: Google style uses 2 spaces instead of 4 spaces in Python code.

File: word_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+
+def sigmoid(x):
+    """
+    Computes the sigmoid function for a scalar
+    x: the scalar
+    """
+    return 1.0/(1.0+np.exp(-x))
+
+
+def analyze_grapheme_clusters(ratios, thrsh):
```
Comment (sffc) [inline word_segmentation.py:45]: This function doesn't return anything; it just prints to the console.  The verb "analyze" is not precise enough.  I would name the function "print_grapheme_clusters" or "plot_grapheme_clusters" to emphasize what the function is actually doing.

File: word_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+    Given break points for words and grapheme clusters, returns the matrix that represents BIES.
+    The output is a matrix of size (n * 4) where n is the number of grapheme clusters in the string
+    char_brkpoints: break points for grapheme clusters
+    word_brkpoints: break points for words
+    """
+    bies = np.zeros(shape=[4, len(char_brkpoints)-1])
+    word_ind = 0
+    for i in range(len(char_brkpoints)-1):
+        word_st = word_brkpoints[word_ind]
+        word_fn = word_brkpoints[word_ind + 1]
```
Comment (sffc) [inline word_segmentation.py:88]: I don't want you to change it in this file, but in the future, consider using "start" and "limit" or even just "0" and "1" for begin and end variable names.  I assume "st" means "start" and "fn" means "finish", but when I see "fn" it looks more like "function" to me.

File: word_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+
+    # Remove texts between following tags
+    line = remove_tags(line, "<NE>", "</NE>")
+    line = remove_tags(line, "<AB>", "</AB>")
+
+    # Remove lines that are all fully in English
+    if is_all_english(line):
+        return -1
+
+    # Exclude cases such as "", " ", "| ", " |", etc.
```
Comment (sffc) [inline word_segmentation.py:184]: Doesn't `is_all_english` catch those cases?

File: word_segmentation.py [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+                unsegmented_line = line.replace("|", "")
+
+                # Making the grapheme clusters brkpoints
+                chars_break_iterator = BreakIterator.createCharacterInstance(Locale.getUS())
+                chars_break_iterator.setText(unsegmented_line)
+                char_brkpoints = [0]
+                for brkpoint in chars_break_iterator:
+                    char_brkpoints.append(brkpoint)
+
+                # Storing the grapheme clusters and their frequency in the dictionary
```
Comment (sffc) [inline word_segmentation.py:240]: You can use a Python Counter for this.

https://docs.python.org/3.8/library/collections.html#collections.Counter

File: word_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+                true_bies = get_bies(char_brkpoints, word_brkpoints)
+                true_bies_str = get_bies_string_from_softmax(np.transpose(true_bies))
+
+                # Compute segmentations of icu and BIES associated with it
+                words_break_iterator = BreakIterator.createWordInstance(Locale.getUS())
+                words_break_iterator.setText(unsegmented_line)
+                icu_word_brkpoints = [0]
+                for brkpoint in words_break_iterator:
+                    icu_word_brkpoints.append(brkpoint)
+                icu_word_segmented_str = get_segmented_string(unsegmented_line, icu_word_brkpoints)
```
Comment (sffc) [inline word_segmentation.py:254]: You don't use `icu_word_segmented_str`

File: word_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+                icu_bies = get_bies(char_brkpoints, icu_word_brkpoints)
+                icu_bies_str = get_bies_string_from_softmax(np.transpose(icu_bies))
+
+                # Counting the number of mismatches between icu_bies and true_bies
```
Comment (sffc) [inline word_segmentation.py:258]: Seems like something to move into a helper function

**Review by sffc — COMMENTED** [review]
Nice!  I got down to about line 388 so far.  Just some style comments.

File: word_segmentation.py (hunk 5 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:88]: The reason that I don't just use "start" and "finish" is that many times I am working with begin and end of more than one thing (like characters and words here). I think the best thing is that I just use a full name that explains what is that variable, such as "word_start" and "word_finish".

File: word_segmentation.py (hunk 6 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:184]: In my original `is_all_english` I excluded spaces and bars before checking if a text is in ASCII. I don't remember why I did that, but I remember I had a reason. For now, I rewrite the `is_all_englihs` function and remove these special cases, hoping it doesn't result in a bug (I didn't detect any by running some of the main functions).

File: word_segmentation.py (hunk 7 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:240]: Yes, if I want to do that I can store all grapheme clusters in a list and then at the end, use `Counter` to get a dictionary from that list. I don't see the point of this though, why I don't make the dictionary directly along the way? This way I don't need to store an extra list too.

File: word_segmentation.py [hunk 10]
```diff
@@ -196,86 +205,84 @@ def clean_line(line):
...
-def preprocess():
+def preprocess_Thai(demonstrate):
     """
     This function uses the BEST data set to
         1) compute the grapheme cluster dictionary that holds the frequency of different grapheme clusters
         2) demonstrate the performance of icu word breakIterator and compute its accuracy
+    Args:
+        demonstrate: shows if we want to see how the algorithm is working or not
     """
+    words_break_iterator = BreakIterator.createWordInstance(Locale.getUS())
```
Comment (sffc) [inline word_segmentation.py:216]: I think in this situation you probably want `Locale.getRoot()` instead of `Locale.getUS()`

File: word_segmentation.py [hunk 11]
```diff
@@ -196,86 +205,84 @@ def clean_line(line):
...
 
 
-def preprocess():
+def preprocess_Thai(demonstrate):
     """
     This function uses the BEST data set to
         1) compute the grapheme cluster dictionary that holds the frequency of different grapheme clusters
         2) demonstrate the performance of icu word breakIterator and compute its accuracy
+    Args:
+        demonstrate: shows if we want to see how the algorithm is working or not
```
Comment (sffc) [inline word_segmentation.py:214]: Optional: this parameter is normally called `verbose`

File: word_segmentation.py [hunk 12]
```diff
@@ -286,35 +293,287 @@ def preprocess():
     return graph_clust_ratio, icu_accuracy
 
 
-def get_BEST_text(starting_text, ending_text):
+def preprocess_Burmese(demonstrate):
+    """
+    This function uses the Okell data set to
```
Comment (sffc) [inline word_segmentation.py:298]: Are you still using Okell?

File: word_segmentation.py [hunk 13]
```diff
@@ -286,35 +293,287 @@ def preprocess():
...
+                            new_line = new_line + "|" + ch
+                        if line[i-1] == "|" and line[i+1] == "|":
+                            new_line = new_line + ch
+                else:
+                    new_line += ch
+            new_line += "\n"
+            wfile.write(new_line)
+
+
+def compute_ICU_accuracy(filename):
```
Comment (sffc) [inline word_segmentation.py:401]: I think you have the ICU BIES code at least 3 times in this file: in preprocess_Thai, preprocess_Burmese, and here in compute_ICU_accuracy.  Consider refactoring the code to reduce duplication.

File: word_segmentation.py [hunk 14]
```diff
@@ -286,35 +293,287 @@ def preprocess():
...
+    train_ratio = 0.4
+    valid_ratio = 0.4
+    train_file = open(train_text, 'w')
+    valid_file = open(valid_text, 'w')
+    test_file = open(test_text, 'w')
+    num_lines = sum(1 for _line in open(input_text))
+    line_counter = 0
+    with open(input_text) as f:
+        for line in f:
+            line_counter += 1
```
Comment (sffc) [inline word_segmentation.py:472]: Hmm.  Some of the data files have high locality, in the sense that there might be more news articles toward the top and religious texts toward the bottom.  I would be more comfortable if we could randomize this across the whole file, or shuffle the lines before bucketing them.

File: word_segmentation.py [hunk 15]
```diff
@@ -0,0 +1,1252 @@
...
+        graph_clust_ids: a dictionary that stores maps from grapheme clusters to integers
+    """
+    # Finding word breakpoints
+    word_brkpoints = []
+    found_bars = 0
+    for i in range(len(input_line)):
+        if input_line[i] == '|':
+            word_brkpoints.append(i - found_bars)
+            found_bars += 1
+    unsegmented_line = input_line.replace("|", "")
```
Comment (sffc) [inline word_segmentation.py:596]: You do a lot of conversion between the segmented and unsegmented versions of text.  You may want to consider bring this distinction into the type system / make it more object-oriented.  For example, have different classes for segmented and unsegmented strings, with conversion methods to go between them.

File: word_segmentation.py (hunk 10 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:216]: Only for this instance, or generally for all `BreakIterator.createWordInstance` and `BreakIterator.createCharacterInstance`?

File: word_segmentation.py (hunk 14 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:472]: Good point! However, I think by randomizing all lines we may lose context that the LSTM model may be able to learn. What I do is that I put every `bucket_size = 20` lines into one group, and then write the first 8 lines in `train_file`, the next 8 lines in `valid_file`, and the remaining in the `test_file`.

File: word_segmentation.py (hunk 7 above)
Comment (sffc) [reply word_segmentation.py]: Another option is `defaultdict`.

https://docs.python.org/3.10/library/collections.html#collections.defaultdict

File: word_segmentation.py [hunk 16]
```diff
@@ -707,6 +728,220 @@ def perfom_bayesian_optimization(hunits_lower, hunits_upper, embedding_dim_lower
     print(optimizer.res)
 
 
+def convert_dic_to_json(dic):
```
Comment (sffc) [inline word_segmentation.py:731]: Better if you `import json`: https://docs.python.org/3/library/json.html

File: word_segmentation.py [hunk 17]
```diff
@@ -707,6 +728,220 @@ def perfom_bayesian_optimization(hunits_lower, hunits_upper, embedding_dim_lower
...
+            self.compute_icu_segmented()
+            self.icu_word_brkpoints = self.compute_word_brkpoints(self.icu_segmented)
+
+        chars_break_iterator = BreakIterator.createCharacterInstance(Locale.getRoot())
+        chars_break_iterator.setText(self.unsegmented)
+        self.char_brkpoints = [0]
+        for brkpoint in chars_break_iterator:
+            self.char_brkpoints.append(brkpoint)
+
+    def compute_icu_segmented(self):
```
Comment (sffc) [inline word_segmentation.py:894]: Here and below: This function looks private.  The naming convention in the Google style guide is to prefix private methods with `_`, like `def _compute_icu_segmented(self):`

File: word_segmentation.py [hunk 18]
```diff
@@ -205,140 +294,115 @@ def clean_line(line):
...
 
                     # Counting the number of mismatches between icu_bies and true_bies
                     icu_total_bies_lengths += len(icu_bies_str)
                     icu_mismatch += diff_strings(true_bies_str, icu_bies_str)
 
+                    # Computing the F1-score between icu_bies and true_bies
+                    f1_output = compute_f1_score(true_bies=true_bies_str, est_bies=icu_bies_str)
+                    correctly_segmented_words += f1_output[1]
+                    true_words += f1_output[2]
+                    segmented_words += f1_output[3]
```
Comment (sffc) [inline word_segmentation.py:345]: If the F1 score is a structured output, you should return a type like a NamedTuple

File: word_segmentation.py [hunk 19]
```diff
@@ -205,140 +294,115 @@ def clean_line(line):
...
                     line_counter += 1
 
-    icu_accuracy = 1 - icu_mismatch/icu_total_bies_lengths
+    icu_bies_accuracy = 1 - icu_mismatch/icu_total_bies_lengths
+
+    precision = correctly_segmented_words / segmented_words
+    recall = correctly_segmented_words / true_words
+    icu_f1_accuracy = 0
+    if precision + recall != 0:
+        icu_f1_accuracy = 2 * precision * recall / (precision + recall)
```
Comment (sffc) [inline word_segmentation.py:362]: I'm not super happy about having the F1 score computation logic in two different places (here and in the main compute_f1_score)

File: word_segmentation.py [hunk 20]
```diff
@@ -205,140 +294,115 @@ def clean_line(line):
     return line
 
 
-def preprocess_Thai(demonstrate):
+################# functions for processing texts #################
+
+def preprocess_thai(verbose):
```
Comment (sffc) [inline word_segmentation.py:299]: A better design would be a function that returns a list of `Line`s based on BEST data, and a separate function that computes F1 score, grapheme cluster dictionary, etc., given the list of `Line`s.

File: word_segmentation.py [hunk 21]
```diff
@@ -349,233 +413,225 @@ def preprocess_Burmese(demonstrate):
     return graph_clust_ratio
 
 
-def add_additional_bars(read_filename, write_filename):
+def get_bies(char_brkpoints, word_brkpoints):
```
Comment (sffc) [inline word_segmentation.py:416]: Could this be a method on `Line`?

File: word_segmentation.py [hunk 22]
```diff
@@ -349,233 +413,225 @@ def preprocess_Burmese(demonstrate):
...
+            continue
+        if char_st != word_st and char_fn == word_fn:
+            bies[2, i] = 1
+            word_ind += 1
+            continue
+        if char_st == word_st and char_fn == word_fn:
+            bies[3, i] = 1
+            word_ind += 1
+            continue
+    return bies
```
Comment (sffc) [inline word_segmentation.py:445]: Consider making BIES a type, which could be a NamedTuple with only one element, a string.  You get better type safety that way.

File: word_segmentation.py [hunk 23]
```diff
@@ -349,233 +413,225 @@ def preprocess_Burmese(demonstrate):
...
+            out += "i"
+        elif mat[i, 2] == max_softmax:
+            out += "e"
+        elif mat[i, 3] == max_softmax:
+            out += "s"
+    return out
 
 
-def divide_train_test_data(input_text, train_text, valid_text, test_text):
+def get_segmented_string_from_bies(line, bies):
```
Comment (sffc) [inline word_segmentation.py:470]: Seems like this could be a method on `Line` perhaps?

File: word_segmentation.py [hunk 24]
```diff
@@ -349,233 +413,225 @@ def preprocess_Burmese(demonstrate):
...
+        precision = correctly_segmented_words / segmented_words
+        recall = correctly_segmented_words / true_words
+        icu_f1_accuracy = 0
+        if precision + recall != 0:
+            icu_f1_accuracy = 2 * precision * recall / (precision + recall)
+        return icu_bies_accuracy, icu_f1_accuracy
 
-def get_file_text(filename):
+
+def normalize_bies(bies_str):
```
Comment (sffc) [inline word_segmentation.py:574]: If you had a first-class type for BIES, you could put functions like this one into that type.

**Review by sffc — CHANGES_REQUESTED** [review]

File: word_segmentation.py [hunk 25]
```diff
@@ -349,233 +413,225 @@ def preprocess_Burmese(demonstrate):
...
-            else:
-                out_str = out_str + " " + line
-            line_counter += 1
-    return out_str
+            elif bies_str[i] == 'b':
+                out_bies += bies_str[i]
+                start_of_word = False
+                continue
+            elif bies_str[i] in ['i', 'e']:
+                if bies_str[i+1] in ['i', 'e']:
```
[the remark this answers is not in the document]
Comment (SahandFarhoodi) [reply word_segmentation.py:599]: Before this `if` I already checked that `i` is not equal to `len(bies_str)`.

File: word_segmentation.py (hunk 16 above)
Comment (SahandFarhoodi) [reply word_segmentation.py:731]: Use `json` only to write down the dictionary? Or to use it for writing down the whole `weigths.json` file?

**Review by sffc — APPROVED** [review]
